src/modules/quick_oeproperties_module.f90

```diff
@@ -268,44 +268,61 @@ subroutine compute_ESP_charge(npoints,xyz_points,esp)
    double precision, external :: rootSquare
 
    integer, allocatable :: IPIV(:)
-   integer :: iatom, jatom, igridpoint, npoints, ierr, NB, LWORK, LDA
-   double precision, intent(in) :: esp(npoints), xyz_points(3,npoints)
+   integer :: iatom, jatom, igridpoint, npoints, ierr, NB, LWORK, LDA, alloc_status
+   double precision, intent(in) :: esp(:), xyz_points(:,:)
    double precision, allocatable :: WORK(:)
    double precision :: A(natom+1,natom+1), B(natom+1), q(natom+1)
    double precision :: distance, distanceb, invdistance, Net_charge
 
-   double precision :: One = 1.0d0, Zero = 0.0d0
+   double precision, allocatable :: invdist_arr(:,:)
 
-!  A and B are initialized. A(natom+1,natom+1) is set to a small number
-!  instead of zero to facilitate diagonalization of A.
+   double precision, parameter :: One = 1.0d0, Zero = 0.0d0
 
-   do iatom = 1, natom
-     B(iatom) = 0
-     do jatom = 1, natom
-       A(jatom,iatom) = 0
-     end do
-   end do
+!  A, B and q are initialized.
+
+   q = Zero
 
+   B = Zero
    B(natom+1) = quick_molspec%molchg
 
-   do jatom = 1, natom
-     A(jatom,natom+1) = 1
-   end do
-   A(natom+1,natom+1) = 0
-
-! The matrix A and vector B is formed.
-
-   do iatom = 1, natom  
-     do igridpoint = 1, npoints
-       distance = rootSquare(xyz(1:3,iatom), xyz_points(1:3,igridpoint), 3)
-       invdistance = 1/distance
-       B(iatom) = B(iatom) + esp(igridpoint) * invdistance
-       do jatom = 1, iatom
-         distanceb = rootSquare(xyz(1:3,jatom), xyz_points(1:3,igridpoint), 3)
-         A(jatom,iatom) = A(jatom,iatom) + invdistance/distanceb
+   A = Zero
+   A(1:natom,natom+1) = One
+
+   allocate(invdist_arr(natom,npoints), stat=alloc_status)
+
+   if (alloc_status /= 0) then
+     ! The matrix A and vector B is formed.
+     do iatom = 1, natom  
+       do igridpoint = 1, npoints
+         distance = rootSquare(xyz(1:3,iatom), xyz_points(1:3,igridpoint), 3)
+         invdistance = 1/distance
+         B(iatom) = B(iatom) + esp(igridpoint) * invdistance
+         do jatom = 1, iatom
+           distanceb = rootSquare(xyz(1:3,jatom), xyz_points(1:3,igridpoint), 3)
+           A(jatom,iatom) = A(jatom,iatom) + invdistance/distanceb
+         end do
+       end do
+     end do
+   else
+     ! First the inverse distance matrix is formed
+     do iatom = 1, natom
+       do igridpoint = 1, npoints
+         invdist_arr(iatom,igridpoint) = 1/rootSquare(xyz(1:3,iatom), xyz_points(1:3,igridpoint), 3)
        end do
      end do
-   end do
+
+     ! Using the inverse distance matrix to form the matrix A and vector B.
+#if defined CUDA
```

Comment:
This should be enabled also for HIP.
```
#if defined(GPU) || defined(MPIV_GPU)
           call GPU_DGEMM (...)
```
but it will require adding DGEMV to util.fh